resolver/GlobalPass.cc

```diff
@@ -242,8 +242,10 @@ void resolveTypeMembers(core::GlobalState &gs, core::ClassOrModuleRef sym,
                     core::cast_type<core::LambdaParam>(attachedClass.asTypeMemberRef().data(gs)->resultType);
                 ENFORCE(lambdaParam != nullptr);
 
-                lambdaParam->lowerBound = core::Types::bottom();
-                lambdaParam->upperBound = sym.data(gs)->externalType();
+                // TODO(jez) factor this logic into shared helper
+                // Allow treating `T.attached_class` as the class itself in final classes
+                lambdaParam->lowerBound = sym.data(gs)->flags.isFinal ? externalType : core::Types::bottom();
+                lambdaParam->upperBound = externalType;
```

Comment:
We can go even further here: in a final class, we can treat the type member as `invariant`, too (I think).